firefox-ios/Client/Frontend/Browser/WebView/PullRefreshView.swift

```diff
@@ -0,0 +1,323 @@
+                // This check prevents progressView re blink when scrolling the pull refresh before the web view is loaded
+                self?.restoreBackgroundProgressViewIfNeeded()
+                let rotationAngle = -(scrollView.contentOffset.y / (self?.frame.height ?? 0.0)) * .pi * 2
+                UIView.animate(withDuration: 0.1) {
+                    self?.progressView.transform = CGAffineTransform(rotationAngle: rotationAngle * 2.0)
+                }
+            }
+        }
+    }
+
+    private func triggerReloadAnimation() {
+        UIView.animate(withDuration: 0.1,
+                       delay: 0,
+                       options: .curveEaseOut,
+                       animations: {
+            self.progressContainerView.transform = UX.progressViewAnimatedBackgroundFinalAnimationTransform
+        }, completion: { _ in
+            self.progressContainerView.backgroundColor = .clear
+            self.progressContainerView.transform = .identity
+            self.progressView.transform = .identity
+            self.onRefreshCallback()
+        })
+    }
+
+    private func blinkBackgroundProgressViewIfNeeded() {
+        guard !refreshIconHasFocus else { return }
+        refreshIconHasFocus = true
+        let shrinkFactor = computeShrinkingFactor()
+        UIView.animate(withDuration: 0.3,
+                       delay: 0,
+                       usingSpringWithDamping: 0.6,
+                       initialSpringVelocity: 10,
+                       animations: {
+            self.progressContainerView.transform = UX.progressViewAnimatedBackgroundBlinkTransform.scaledBy(x: shrinkFactor,
+                                                                                                            y: shrinkFactor)
+            self.progressContainerView.backgroundColor = self.currentTheme?.colors.layer4
+        }, completion: { _ in
+            UIImpactFeedbackGenerator(style: .heavy).impactOccurred()
+        })
+    }
+
+    private func restoreBackgroundProgressViewIfNeeded() {
+        guard refreshIconHasFocus else { return }
+        refreshIconHasFocus = false
+        UIView.animate(withDuration: 0.3) {
+            self.progressContainerView.transform = .identity
+            self.progressContainerView.backgroundColor = .clear
+        }
+    }
+
+    private func showEasterEgg() {
+        guard let easterEggGif else { return }
+        easterEggGif.isHidden = false
+        let angle = atan2(easterEggGif.transform.b, easterEggGif.transform.a)
+        UIView.animate(withDuration: 0.3) {
+            easterEggGif.transform = .identity.rotated(by: angle)
+        }
+    }
+
+    private func computeShrinkingFactor() -> CGFloat {
+        guard let scrollView else { return 1.0 }
+        let minDimension = min(scrollView.frame.height, scrollView.frame.width)
+        if minDimension / 4.0 > UX.blinkProgressViewStandardThreshold {
+            return 1
+        }
+        return 0.8
+    }
```

Comment:
I'd love to have a little documentation for this function so I can understand better 😃

Reply:
Yes sure 🚀

Reply:
Excellent, thank you, this makes sense for me! 👍